utils_linux.go

```diff
@@ -219,7 +220,12 @@ func createPidFile(path string, process *libcontainer.Process) error {
 
 // XXX: Currently we autodetect rootless mode.
 func isRootless() bool {
-	return os.Geteuid() != 0
+	// Even if os.Geteuid() == 0, it might still require rootless mode,
+	// especially when running within `unshare -m -r`.
+	// So we use system.GetParentNSeuid() here.
+	//
+	// TODO(AkihiroSuda): how to support nested userns?
```

Comment:
Rather than checking UID, it might be better to attempt some lightweight operation that requires "real" root privilege.

Any thought?

Reply:
The simplest test would be to do `mknod`, but I'm not sure I think this is the best idea in the world to be honest. There are plans in the future to provide the ability for previously disallowed syscalls to be allowed through the whole SECCOMP syscall emulation mechanism that's being worked on, so trying to pin down "we are in a user namespace because X doesn't work" is a bit of a stretch.

Reply:
`func isRootless() { return os.Geteuid() != 0 || readFile("/proc/self/setgroups") == "deny") }` is fine then?